pkg/containertools/runner.go

```diff
@@ -137,18 +138,49 @@ func (r *ContainerCommandRunner) Unpack(image, src, dst string) error {
 	}
 
 	id := strings.TrimSuffix(string(out), "\n")
-	args = r.argsForCmd("cp", id+":"+src, dst)
+
+	// Create a new exporter to copy and read output from stdout to a tar reader.
+	// This is done to rewrite file permissions on the fly to avoid permissions-related errors in standard docker cp.
+	exporter, err := NewExporter(dst, r.logger)
+	if err != nil {
+		return fmt.Errorf("setting unpacking destination: #%v", err)
+	}
+	// write files to stdout and then read tar stream
+	args = r.argsForCmd("cp", id+":"+src, "-")
 	command = exec.Command(r.containerTool.String(), args...)
+	command.Stdout = exporter.Writer()
 
 	r.logger.Infof("running %s cp", r.containerTool)
 	r.logger.Debugf("%s", command.Args)
 
-	out, err = command.CombinedOutput()
+	// setup exporter
+	var wg sync.WaitGroup
+	var expError error
+	wg.Add(1)
+	go func(wg *sync.WaitGroup) {
+		defer wg.Done()
+		err := exporter.Run()
+		if err != nil {
+			expError = fmt.Errorf("exporting container filesystem as a tar stream: %v", err)
+		}
+	}(&wg)
+
+	err = command.Run()
 	if err != nil {
-		r.logger.Errorf(string(out))
+		exporter.Close()
 		return fmt.Errorf("error copying container directory %s: %v", string(out), err)
 	}
 
+	if expError != nil {
+		exporter.Close()
+		return expError
+	}
+
+	if err := exporter.Close(); err != nil {
```

Comment:
so calling `Close()` sends the EOF signal on the writer side of the pipe, which causes the tar reader to stop reading, which results in the goroutine exiting. 

`wg.Wait()` without the `Close()` first causes the test case to hang, as the read and write pipes don't close properly, from what I've observed. Maybe it's because I'm using a map with dummy files as a data structure versus a proper tar archive though.